### PR TITLE
Jira organizational components [#147302081]

### PR DESCRIPTION
**Pull request by nc-firthh**
https://www.pivotaltracker.com/story/show/147302081

### CODE REVIEW

**Comment by nc-firthh** [comment]
This was my first time doing any sort of "from scratch" react app design. So if you have any feedback, criticisms, nitpicks or whatever else in regards to my design choices, I'd be really interested in hearing them.

Thanks!

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,4 @@
 build/*
 node_modules/*
 .DS_Store
-pivotal_token.js
+token.js
```
Comment (vsarang) [inline .gitignore]: let's call this `jira_token.js` so it's really clear

File: gulpfile.babel.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 import gulp from 'gulp';
+import server from 'gulp-express';
```
Comment (vsarang) [inline gulpfile.babel.js]: np: imports should be ordered lexicographically by import path

File: server.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+const _ = require('lodash');
+var express = require('express');
+var app = module.exports.app = exports.app = express();
```
Comment (vsarang) [inline server.js]: np: `module.exports` should be at the bottom of the file

File: server.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+const _ = require('lodash');
```
Comment (vsarang) [inline server.js]: Let's break these imports up into their natural groups. In DO we've been doing
```
external dependencies
internal dependencies (e.g. submodules)
internals
```
and the internals can be further broken up into constants, services, components, etc.

File: server.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  port: 443,
+  method: 'GET',
+  headers: {
+    'Content-Type': 'application/json',
+    'Authorization': 'Basic '+ TOKEN
+  }
+};
+
+var get = function (options, cb)
+{
```
Comment (vsarang) [inline server.js]: function syntax should be as follows:
```js
function name(arg1, arg2) {
}
```
Note that the open bracket `{` falls on the same line, and there is no space between the function name and the parenthesis -- this is to differentiate functions from control statements like
```js
if (statement) {
}
```

File: server.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  var prot = options.port == 443 ? https : http;
```
Comment (vsarang) [inline server.js]: Should this be `port`? If not, `prot` is not a good variable name -- I have no idea what it means.

File: server.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  var req = prot.request(options, function(res)
```
Comment (vsarang) [inline server.js]: Let's just call this `request`, and the argument `response`. It's a lot easier to read without abbreviations, and it's not like the full names are very long. The close we can get our code to plain English, the better.

Comment (vsarang) [inline server.js]: Same comment here about the function by the way, the `{` should fall on the same line.

File: server.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    res.on('data', function (chunk) {
+      output += chunk;
+    });
+
+    res.on('end', function() {
+      cb(res.statusCode, output);
+    });
+  });
+
+  req.on('error', function(err) {
```
Comment (vsarang) [inline server.js]: Do we need an empty function declaration here? Seems like we can nix this.

File: server.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+var get = function (options, cb)
+{
+  var prot = options.port == 443 ? https : http;
+  var req = prot.request(options, function(res)
+  {
+    var output = '';
+    res.setEncoding('utf8');
+
+    res.on('data', function (chunk) {
```
Comment (vsarang) [inline server.js]: Consistent style is very important, please go through and fix all the function declarations.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    });
+  });
+
+  req.on('error', function(err) {
+  });
+
+  req.end();
+};
+
+app.use(express.static('.'));
```
Comment (vsarang) [inline server.js]: Let's put some comments in here about what is going on, this is really hard to follow.

File: server.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+ */
+app.get('/history', (request, response) => {
+  let uri = `/rest/api/2/issue/${request.query.key}?expand=changelog&maxResults=100`;
+  options.path = encodeURI(uri);
+  get(options, ({ statusCode, body }) => {
+    response.status(statusCode).send(body);
+  });
+});
+
+app.listen(1337);
```
Comment (vsarang) [inline server.js]: Let's extract config like this out to constants and put them at the top of the file.
```js
const PORT = 1337; // elite h4x0rz
```

File: src/scripts/components/bio.jsx [hunk 12]
```diff
@@ -1,15 +1,15 @@
...
     return (
-      <div className="bb-r-container-content-block bio-column">
+      <div className='bb-r-container-content-block bio-column'>
         {
           _.map(this.props.bio, (par, i) => (
               <div
-                className="bio-paragraph"
-                key={ "par-" + i }
+                className='bio-paragraph'
+                key={ 'par-' + i }
```
Comment (vsarang) [inline src/scripts/components/bio.jsx]: This can be a template string.

File: src/scripts/components/jira/JiraFooter.jsx [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import _ from 'lodash';
+import React from 'react';
+import moment from 'moment';
```
Comment (vsarang) [inline src/scripts/components/jira/JiraFooter.jsx]: np: imports should be ordered alphabetically by import path

File: src/scripts/components/jira/JiraFooter.jsx [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import PropTypes from 'prop-types';
+
+const JiraFooter = (props) => {
+  const dotActivity = _.times(props.numViews, (index) => {
```
Comment (vsarang) [inline src/scripts/components/jira/JiraFooter.jsx]: What is this for?

File: src/scripts/components/jira/JiraHeader.jsx [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
```
Comment (vsarang) [inline src/scripts/components/jira/JiraHeader.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+const MonthlyGoalsBody = (props) => (
+  <div className='monthly-goals-body'>
+    <MonthlyGoalsTable
+      pageIndex={ props.currentPage }
+      stories={ props.stories }
+    />
+    {
+      props.stories.length > 1 && (
+      <MonthlyGoalsFooter
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx]: np: tab this over once more

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+  </div>
+);
+
+MonthlyGoalsBody.propTypes = {
+  currentPage: PropTypes.number,
+  daysRemaining: PropTypes.number,
+  stories: PropTypes.array
+};
+
+export default MonthlyGoalsBody;
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsBody.jsx]: Just noticed this, but you might want to look into configuring your text editor to insert the correct newline character at the ends of your files. That's a POSIX standard that we use as a convention.

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsFooter.jsx [hunk 19]
```diff
@@ -0,0 +1,29 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsFooter.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsRow.jsx [hunk 20]
```diff
@@ -0,0 +1,44 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsRow.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsRow.jsx [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import React from 'react';
+import PropTypes from 'prop-types';
+import config from './config'
+
+import MonthlyGoalsStatusIndicator from './MonthlyGoalsStatusIndicator';
+
+const MonthlyGoalsRow = (props) => {
+  const done = props.status.statusCategory.key === 'done' ? ' done' : '';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsRow.jsx]: looks like you could be leveraging the `classNames` helper here to better communicate the intentions of this code

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsStatusIndicator.jsx [hunk 22]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsStatusIndicator.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsTable.jsx [hunk 23]
```diff
@@ -0,0 +1,47 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsTable.jsx]: np: import order

File: src/scripts/components/jira/monthly-goals/MonthlyGoalsTable.jsx [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+        _.map(props.stories[props.pageIndex], (story, index) => (
+          <MonthlyGoalsRow
+            key={ story.id }
+            category={ story.fields.issuetype }
+            dateInSit={ story.dateInSit }
+            isHighlight={ story.addDate && story.addDate.getMonth() === new Date().getMonth() }
+            issue={ story.key }
+            name={ story.fields.summary }
+            points={ story.fields.customfield_10117 }
+            rank={ props.pageIndex*props.stories[0].length + index + 1 }
```
Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsTable.jsx]: np: pad mathematical operations `*`

Comment (vsarang) [inline src/scripts/components/jira/monthly-goals/MonthlyGoalsTable.jsx]: There might be an eslint rule for this

File: src/scripts/components/jira/overall-progress/OverallProgressBody.jsx [hunk 25]
```diff
@@ -0,0 +1,42 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressBody.jsx]: np: import order

File: src/scripts/components/jira/overall-progress/OverallProgressBuckets.jsx [hunk 26]
```diff
@@ -0,0 +1,24 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressBuckets.jsx]: np: import order (I'm gonna stop commenting on all of these)

File: src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+
+import config from './config';
+
+const OverallProgressHighlight = (props) => {
+  const newSprint = _.find(props.change.items, { fieldId : 'customfield_10113' });
+  const newStatus = _.find(props.change.items, { fieldId : 'status' });
+
+  const hasBeenAdded = newSprint && newSprint.to == props.currentSprintId;
+  const changeText = newSprint ?
+    hasBeenAdded ? config.sprint.messages.addedText : config.sprint.messages.removedText :
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx]: Let's not chain ternaries, they are really hard to parse.

File: src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+    newStatus && config.status.messages[props.statusCategory.key]
+
+  return (
+    <div className='overall-progress-highlight'>
+      <div className='description'>
+        <span className='issue'> { props.issue }</span>
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx]: extra space before the`{` on some of these lines

File: src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+        <span className='summary'> { props.summary }</span>
+        <span className='change'>
+          <span className='transition'> { changeText.transition }</span>
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressHighlight.jsx]: let's get newlines on these two spans (passing a prop and children makes it hard to parse)

File: src/scripts/components/jira/overall-progress/OverallProgressMonths.jsx [hunk 30]
```diff
@@ -0,0 +1,30 @@
+import _ from 'lodash';
+import moment from 'moment';
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const OverallProgressMonths = (props) => (
+  <div className='overall-progress-months'>
+    {
+      _.map(props.featureCounts, (featureCount, index) => (
+        <div key={ index } className={ 'current-sprint' + (index === 0 ? ' active' : '') }>
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressMonths.jsx]: np: multiple props need multiple lines

File: src/scripts/components/jira/overall-progress/OverallProgressMonths.jsx [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+          <div className='sprint-summary'>
+            <div className='month-label'>
+              { moment().add(index, 'months').format('MMMM') }
+            </div>
+            <div className='feature-count'>
+              { featureCount + ' Issues Queued' }
```
Comment (vsarang) [inline src/scripts/components/jira/overall-progress/OverallProgressMonths.jsx]: np: template string

File: src/scripts/components/shared/LoadingPage.jsx [hunk 32]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+
+const LoadingPage = () => (
+  <div className="loading-page">
+    <div className="loading-animation">
+      <label> ●</label>
```
Comment (vsarang) [inline src/scripts/components/shared/LoadingPage.jsx]: np: there's an extra space here and on the lines below

File: src/scripts/slides/jira_slide.jsx [hunk 33]
```diff
@@ -0,0 +1,281 @@
...
+import JiraFooter from '../components/jira/JiraFooter'
+import JiraHeader from '../components/jira/JiraHeader'
+import LoadingPage from '../components/shared/LoadingPage'
+import MonthlyGoalsBody from '../components/jira/monthly-goals/MonthlyGoalsBody';
+import OverallProgressBody from '../components/jira/overall-progress/OverallProgressBody';
+
+const BOARD_ID = 4;
+
+// Polyfill to give promises to the TV
+const Promise = require('es6-promise').Promise;
```
Comment (vsarang) [inline src/scripts/slides/jira_slide.jsx]: We can't `import` this, I'm guessing?

File: src/scripts/slides/jira_slide.jsx [hunk 34]
```diff
@@ -0,0 +1,281 @@
...
+
+let pageInterval;
```
Comment (vsarang) [inline src/scripts/slides/jira_slide.jsx]: np: `let pageInterval, viewInterval;`

File: src/scripts/slides/jira_slide.jsx [hunk 35]
```diff
@@ -0,0 +1,281 @@
...
+  },
+
+  changePage() {
+    const nextPage = this.state.stories &&
+      this.state.stories.length - 1 === this.state.currentPage ? 0 : this.state.currentPage + 1;
+    this.setState({ currentPage: nextPage });
+  },
+
+  changeView() {
+    const nextView = this.state.view === 'overall' ? 'monthly' : 'overall';
```
Comment (vsarang) [inline src/scripts/slides/jira_slide.jsx]: np: extract to constants

File: src/scripts/components/jira/JiraFooter.jsx (hunk 14 above)
Comment (nc-firthh) [inline src/scripts/components/jira/JiraFooter.jsx]: `dotActivity` is for the view indicator dots in the footer. I wanted an array of booleans corresponding to which views were active and which inactive so I could have a convenient way of rendering those indicator dots and applying the appropriate classnames.

Comment (vsarang) [inline src/scripts/components/jira/JiraFooter.jsx]: The PageIndicator should probably be extracted to its own component.

Comment (nc-firthh) [inline src/scripts/components/jira/JiraFooter.jsx]: Good point, I like that.

File: server.js [hunk 36]
```diff
@@ -0,0 +1,97 @@
+import _ from 'lodash';
```
Comment (vsarang) [inline server.js]: np: import order (I know `_` should technically come first, but we order by import _path_, so `lodash` comes after `express` and `http`/`https`.

File: server.js [hunk 37]
```diff
@@ -0,0 +1,97 @@
...
+const options = {
+  host: 'nextcapital.atlassian.net',
+  port: 443,
+  method: 'GET',
+  headers: {
+    'Content-Type': 'application/json',
+    'Authorization': 'Basic ' + JIRA_TOKEN
+  }
+};
+const PORT = 1337; // elite h4x0rz
```
Comment (vsarang) [inline server.js]: np: constants should be at the _very_ top of the file (after imports)

💯  on the comment btw

File: server.js [hunk 38]
```diff
@@ -0,0 +1,97 @@
...
+
+function getOptions(path) {
+  return _.assign(options, { path: path });
```
Comment (vsarang) [inline server.js]: So when you pass `options` as the first argument into `_.assign`, it actually gets mutated. In order to preserve the original object, we have to do `_.assign({}, options, { path });` (also note the shorthand we can use when the name of the property matches the name of the variable we are assigning to it).

File: server.js [hunk 39]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+function getQueryString(query) {
+  return _.reduce(_.omit(query, 'sprint'), (result, value, key) => {
+    return _.concat(result, `${key}=${value}`);
+  }, []).join('&');
+}
+
+function httpGet(options, callback) {
+  const port = options.port === 443 ? https : http;
```
Comment (vsarang) [inline server.js]: Oh you know what... it wasn't a typo before... I think `prot` was short for `protocol`, not a typo of `port`. Now you know why I like using the full name for variables instead of abbreviations. Let's fix this and update the variable name.

File: src/scripts/components/jira/JiraFooter.jsx [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+    return index === props.pageIndex;
+  });
+
+  return (
+    <div className='jira-footer'>
+      <div className='logo'>
+        <img src='images/nc-logo-white-full.svg' />
+      </div>
+      <div className='page-indicator'>
```
Comment (vsarang) [inline src/scripts/components/jira/JiraFooter.jsx]: When I said extract to its own component, I meant this should be its own full React component --
```jsx
<PageIndicator
  numViews={ props.numViews }
  activeView={ props.pageIndex }
/>
```
Seems like it's being used in the `MonthlyGoalsFooter` as well, so let's be sure to consume it there too.

File: server.js (hunk 37 above)
Comment (nc-firthh) [inline server.js]: So the sequencing here is uppercase constants first, followed by normal, camelCase constants like `app` and `options`?

Comment (vsarang) [inline server.js]: Well these are all `const`s, but in programming terms their use cases are very different. `app` is the instance of our server. The _reference_ does not change (it is always one object), but it is mutable, so it's very different from `options` and `PORT`. Those two variables will never change -- their references will not change, and their substance will not change either. I would argue `options` should be renamed to `REQUEST_OPTIONS` to reflect how you are using it (it's never being used to configure the server, but rather the HTTP requests it makes). With all that in mind, this is how I would structure it:

```js
const PORT = 1337;
const REQUEST_OPTIONS = {
  // ...
};

function getOptions(path) {
  // ...
}
// ... other functions

const app = express();
app.use(express.static('.'));
// ... other app initialization

// finally, start the server
app.listen(SERVER_PORT);
export default app;
```

I would also consider renaming `app` to `server`, since that's what it really is.

Comment (vsarang) [inline server.js]: Also, is anything ever importing `app`? If not, why are we exporting it?

Comment (nc-firthh) [inline server.js]: Cool, thanks for the detailed answer! I agree with all of those suggestions.

And good catch on the un-used export. It was a hold-over from an earlier version of the server, and I never thought to check whether it was actually needed. I'm not whether it was ever being used at one point, but regardless, it doesn't appear that it's being used now, and everything seems to work fine without it, so I think it makes sense to take it out.
